wellcomecollection/wellcomecollection.org
# Handle Segment script behaviour on consent change

## Who is this for?
#10964

## What is it doing for them?
- Moves the Segment script to the body in `_app` and out of `_document` so we could change it based on client side changes and state.
- Adds analytics consent status to state to affect the above when needed.
- Adds `load: false` value ([docs](https://segment.com/docs/connections/sources/catalog/libraries/website/javascript/#load)) to Segment script's options if consent has not been granted. The script doesn't disappear, but it's load function gets removed/not triggered by default.
<img width="330" alt="Screenshot 2024-07-04 at 12 28 17" src="https://github.com/wellcomecollection/wellcomecollection.org/assets/110461050/0103b66e-e651-4ad2-8717-f01af736d5ef"> <img width="330" alt="Screenshot 2024-07-04 at 12 28 22" src="https://github.com/wellcomecollection/wellcomecollection.org/assets/110461050/353a574d-e64f-4aae-a489-8829b71ba5a7">
- With `useEffect`, re-renders the script and its values (see above) should the state of the analytics consent change. 
- Rename event that gets triggered on `onAccept` and `onRevoke` from `analyticsConsentChanged` to `consentChanged`, as we have marketing as well now. This was a miss from when we added a new category and made things confusing.
- Added the analytics consent state to the `trackPageview` useEffect's dependency array. Now, if someone toggles analytics consent on or off, or Accepts all cookies the first time they see the banner, if there are pageview values, the tracking gets triggered again.

## How to test
- Add `WC_analyticsDebug` to your cookies (value: `true`), which will log events sent to Segment.
- Play around with all scenarios of cookies
- This page is a good one to test with as it has a `pageview` set-up: http://localhost:3000/articles/ZWTLeBUAACgAVku2
- Test going to a 404, it shouldn't break (has been an issue in the past) and cookies should work just like anywhere else.

## Risks
We've misunderstood something and it affects Segment events. So I suggest we merge early next week and keep a close eye.
